crates/core/src/pipeline/runner.rs

```diff
@@ -108,9 +125,144 @@ impl PipelineRunner {
             fps_last_frame_ctr: 0,
             info: None,
             thumb_interval: 1800, // Disable thumbnails by default for performance
+            state: RunnerState::Normal,
         })
     }
 
+    /// Process a single idle frame - generates one source frame and processes it through all variants
+    unsafe fn process_single_idle_frame(&mut self, config: &PipelineConfig) -> Result<()> {
+        use std::time::{Duration, Instant};
+        
+        if config.variants.is_empty() {
+            return Ok(());
+        }
+
+        // Extract timing info from current state
+        let (mut last_frame_time, variant_index) = match &mut self.state {
+            RunnerState::Idle { last_frame_time, variant_index, .. } => (last_frame_time, variant_index),
+            _ => return Ok(()), // Only process in idle state
+        };
+
+        // Time-based frame rate calculation
+        let now = Instant::now();
+        if let Some(last_time) = *last_frame_time {
+            // Calculate target frame interval (assume 30fps for now)
+            let target_interval = Duration::from_millis(33); // ~30fps
+            let elapsed = now.duration_since(last_time);
+            
+            if elapsed < target_interval {
+                // Not time for next frame yet
+                std::thread::sleep(target_interval - elapsed);
+            }
+        }
+        *last_frame_time = Some(Instant::now());
+
+        // Find the primary video variant to determine source frame properties
+        let video_variant = config.variants.iter().find_map(|v| {
```

Comment:
Instead of searching for a video variant just store the demuxer info in the piplineconfig struct during setup_pipeline, then your placeholder generator can take the DemuxInfo stream struct to generate corresponding test data based on the input stream